### PR TITLE
`@wagmi/core` & the platform-agnostic `WagmiClient` concept

### PR DESCRIPTION
**Pull request by jxom**
### Motivation

The idea of this PR is to extract the "core" logic of the `wagmi` context & hooks into an agnostic `@wagmi/core` package that can be used with any React async state lib (react-query, swr, etc), as well as frameworks like Svelte, Vue or Vanilla JS.

This PR is a precursor & requirement to adding react-query support into `wagmi` + supporting a `@wagmi/swr` package

👉 https://github.com/tmm/wagmi/pull/153

### Implementation

This PR:
- Adds a `WagmiClient`, which is essentially the `wagmi` React context converted to Vanilla JS (using Zustand for reactivity) – `packages/core/src/client.ts`
- Converts the internal logic of the existing React hooks into a set of core "actions" that utilise `WagmiClient` – `packages/core/src/actions/*`

### Usage

Usage in its entirety can be seen in the `examples/using-core` folder.

#### Getting started

First, you will need to initialise a "wagmi client" – the arguments are the same as what you'd pass to `WagmiProvider` in `wagmi`:

```tsx
import { createWagmiClient } from '@wagmi/core';

const wagmiClient = createWagmiClient({
  connectors,
  provider,
  webSocketProvider,
})

function App() {
  ...
}
```

Optional: If you wish to enable auto-connection, you will need to do that on initial load:

```tsx
function App() {
  React.useEffect(() => {
    const autoConnect = async () => {
      await wagmiClient.autoConnect()
    }
    autoConnect()
  }, [])
}
```

Then, you can start using the actions like so:

```tsx
import * as React from 'react'
import { getAccount, watchAccount } from '@wagmi/core'

export const Account = () => {
  const [accountData, setAccountData] = React.useState(getAccount())
  React.useEffect(() => {
    const unwatch = watchAccount(setAccountData)
    return unwatch
  }, [])

  if (!accountData) return <div>No account connected</div>
  return (
    <div>
      address: {accountData.address}
    </div>
  )
}
```

#### Actions

Actions are split up into:

##### Read actions

> Note: A `get` or `fetch` prefix denotes if a function is sync or async, respectively.

- [x] `accounts/getAccount` (@jxom)
- [x] `accounts/getNetwork` (@jxom)
- [x] `accounts/fetchBalance` (@jxom)
- [x] `accounts/fetchSigner` (@jxom)
- [x] `contracts/getContract` (@jxom)
- [x] `ens/fetchEnsAddress` (@tmm)
- [x] `ens/fetchEnsAvatar` (@tmm)
- [x] `ens/fetchEnsName` (@tmm)
- [x] `ens/fetchEnsResolver` (@tmm)
- [x] `network-status/fetchBlockNumber` (@sammdec)
- [x] `network-status/fetchFeeData` (@sammdec)
- [x] `tokens/fetchToken` (@sammdec)

##### Watch actions

Watch actions provide reactivity & invoke their callback function (if needed) when the global wagmi store (or external event) updates

- [x] `accounts/watchAccount` (@jxom)
- [x] `accounts/watchBalance` (@jxom)
- [x] `accounts/watchNetwork` (@jxom)
- [x] `accounts/watchSigner` (@jxom)
- [x] `contracts/watchContractEvent` (@jxom)
- [x] `contracts/watchToken`
- [x] `ens/watchEnsAddress` (@tmm)
- [x] `ens/watchEnsAvatar` (@tmm)
- [x] `ens/watchEnsName` (@tmm)
- [x] `ens/watchEnsResolver` (@tmm)
- [x] `network-status/watchBlockNumber`
- [x] `network-status/watchFeeData`

##### Perform actions

- [x] `accounts/connect` (@jxom)
- [x] `accounts/disconnect` (@jxom)
- [x] `accounts/signMessage` (@jxom)
- [x] `contracts/readContract` 
- [x] `contracts/writeContract` (@jxom)
- [x] `transactions/sendTransaction` (@jxom)
- [x] `transactions/waitForTransaction` (@jxom)

### CODE REVIEW

File: packages/core/package.json [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      "optional": true
+    },
+    "walletlink": {
+      "optional": true
+    }
+  },
+  "dependencies": {
+    "@ethersproject/providers": "^5.5.1",
+    "eventemitter3": "^4.0.7",
+    "valtio": "^1.3.0"
```
Comment (jxom) [inline packages/core/package.json:58]: ignore this... experimented w/ proxy state for reactivity, but failed

**Review by sammdec — COMMENTED** [review]
This is awesome! Exactly what i had in my head! Is it in a place where we can start building out the wrappers for react-query etc @jxom ?

**Comment by jxom** [comment]
@sammdec – sure, i was thinking we could just smash out all the core actions (`packages/core/src/actions`), and then perhaps write react-query wrappers after? we could divide and conquer these actions?

@tmm what do you think?

**Comment by jxom** [comment]
cc/ @shunkakinoki

**Comment by sammdec** [comment]
> @sammdec – sure, i was thinking we could just smash out all the @wagmi/core package actions (packages/core/src/actions), and then perhaps write react-query wrappers in the wagmi package after? we could divide and conquer these actions?

Yeah sounds good to me, which ones should I pick up?

**Comment by tmm** [comment]
Let's do it! Only thing is let's try to have `packages/core/src/actions` and `packages/wagmi/src/hooks` have a consistent directory structure.

**Comment by jxom** [comment]
> > @sammdec – sure, i was thinking we could just smash out all the @wagmi/core package actions (packages/core/src/actions), and then perhaps write react-query wrappers in the wagmi package after? we could divide and conquer these actions?
> 
> Yeah sounds good to me, which ones should I pick up?

@sammdec – I have added a checklist to the PR description, maybe just comment which ones you want to take then I can tick them off?

**Comment by tmm** [comment]
We should figure out how we want to document `WagmiClient`too. Maybe we split up the docs into core and react?

**Comment by sammdec** [comment]
@jxom I can take this lot to start

> network-status/fetchBlockNumber
>  network-status/fetchFeeData
>  providers/getProvider
>  providers/getWebSocketProvider
>  tokens/fetchToken

**Comment by jxom** [comment]
Sounds great @sammdec

**Pull request merged**